### PR TITLE
Dynamic properties migration

### PR DESCRIPTION
**Pull request by BeadW**
Follow up to #25, This PR adds to #26 so that we can migrate from any prior set of yaml properties to the 1:1 model -> yaml pattern.

Conversation is already in #25

### CODE REVIEW

File: dbt_invoke/properties.py [hunk 1]
```diff
@@ -268,104 +278,158 @@ def migrate(
...
-            try:
-                property_file_dict = _get_property_header(
-                    resource_name,
-                    data['resource_type'],
-                    data['properties'],
-                )
-                _utils.write_yaml(data['path'], property_file_dict, mode='x')
-                indices_to_remove[data['resource_type_plural']].append(
-                    data['index']
+        if patch_path not in migration_map:
```
Comment (robastel) [inline dbt_invoke/properties.py:336]: I may be missing something obvious, but I am unable to think of a situation where `patch_path` would already be in `migration_map`. Did a situation like that arise during development or testing?

File: dbt_invoke/properties.py [hunk 2]
```diff
@@ -268,104 +278,158 @@ def migrate(
...
+            migration_map[patch_path] = [
+                {
+                    'resource_name': nodes[node]['name'],
+                    'resource_type': nodes[node]['resource_type'],
+                    'resource_type_plural': None if nodes[node]['resource_type'] not in _SUPPORTED_RESOURCE_TYPES else _SUPPORTED_RESOURCE_TYPES[nodes[node]['resource_type']],
```
Comment (robastel) [inline dbt_invoke/properties.py:341]:
```suggestion
                    'resource_type_plural': _SUPPORTED_RESOURCE_TYPES.get(nodes[node]['resource_type']),
```

File: dbt_invoke/properties.py [hunk 3]
```diff
@@ -268,104 +278,158 @@ def migrate(
...
         )
-    else:
-        _LOGGER.info(f'Nothing to migrate from {str(location.resolve())}')
+        # Remove resource types from the source yaml
+        for resource_type_plural, resources in current_property_yaml.items():
+            if resource_type_plural in _SUPPORTED_RESOURCE_TYPES.values() and len(resources) == 0:
+                current_property_yaml.pop(resource_type_plural)
+        # Check if there are any properties left in the source file.
+        # If there aren't then delete the file.
+        if not set(_SUPPORTED_RESOURCE_TYPES.values()).intersection(current_property_yaml.keys()):
```
Comment (robastel) [inline dbt_invoke/properties.py:421]: I think we may often want to avoid deleting the property files for a couple of reasons:
- There are unsupported resource types like sources and exposures that may exist in these property files
- There may be comments in the property files outside of the model definitions that we do not migrate

What do you think about only deleting the property files that contain no non-whitespace characters?

File: dbt_invoke/properties.py [hunk 4]
```diff
@@ -206,19 +208,29 @@ def echo_macro(ctx):
         f'\n{_utils.get_macro(_MACRO_NAME)}'
     )
 
+@functools.lru_cache
```
Comment (robastel) [inline dbt_invoke/properties.py:211]: My knowledge on `@functools.lru_cache` is limited, but is it helping us here given that we only run `_read_manifest` once in each invocation of the `migrate` command?

**Review by robastel — COMMENTED** [review]
Sorry for the delay. I should be able to be more responsive this week.

I have left a few questions for you to help ensure my understanding. Thanks again @BeadW !

File: dbt_invoke/properties.py (hunk 1 above)
Comment (BeadW) [reply dbt_invoke/properties.py:336]: The `patch_path` is the current config file where the resource exists so if there is more than one resource defined in that file then as we are building the `migration_map` we will see the value more than once.

File: dbt_invoke/properties.py (hunk 3 above)
Comment (BeadW) [reply dbt_invoke/properties.py:421]: Yep that is a good point. I will have to take a read on how to determine this is true and make the change.

File: dbt_invoke/properties.py (hunk 4 above)
Comment (BeadW) [reply dbt_invoke/properties.py:211]: I think my prior cut was calling `_read_manifest` more than once so I likely just neglected to remove it. I will add a change for this when I push the commit for deletion of existing yml files.

File: dbt_invoke/properties.py (hunk 3 above)
Comment (robastel) [reply dbt_invoke/properties.py:421]: So I started looking into it and I don't see an officially supported way in `ruamel.yaml`'s documentation (https://yaml.readthedocs.io/en/latest/detail.html?highlight=comments#round-trip-including-comments) to retrieve the comments and formatting. 

What are your thoughts on the following process:
1. Use Path.read_text() to read the contents of the newly overwritten property file as a string
2. `strip()` the string
3. Delete the file if the result of `strip()` is an empty string

File: dbt_invoke/properties.py [hunk 5]
```diff
@@ -268,104 +278,158 @@ def migrate(
...
+                        Path(
+                            Path(
+                                ctx.config['project_path']
+                            ) / 
+                            resources_map[nodes[node]['name']]['original_file_path']).parents[0] / nodes[node]['name']
+                            ).with_suffix('.yml'),
+                    'resource_index': '',
+                    'resource_properties': ''
+                }
+            ]
```
Comment (robastel) [inline dbt_invoke/properties.py:369]:
```suggestion
        if patch_path not in migration_map:
            migration_map[patch_path] = list()
        migration_map[patch_path].append(
            {
                'resource_name': nodes[node]['name'],
                'resource_type': nodes[node]['resource_type'],
                'resource_type_plural': _SUPPORTED_RESOURCE_TYPES.get(nodes[node]['resource_type']),
                'resource_destination': (
                    Path(
                        ctx.config['project_path'],
                        resources_map[nodes[node]['name']]['original_file_path'],
                    ).with_suffix('.yml'),
                    'resource_index': '',
                    'resource_properties': ''
            }
        )
```

File: dbt_invoke/properties.py (hunk 3 above)
Comment (BeadW) [reply dbt_invoke/properties.py:402]:
```suggestion
        if Path(location).read_text().strip() == "":
```

Comment (BeadW) [reply dbt_invoke/properties.py:402]: Seems simple enough. I agree there isn't a great looking way to achieve it based on the interpretation of the yaml. I don't see much of an issue with reading the file again at this end stage for performance. Most of the time it will be empty.

File: dbt_invoke/properties.py (hunk 4 above)
Comment (BeadW) [inline dbt_invoke/properties.py:211]:
```suggestion
```

File: dbt_invoke/properties.py [hunk 6]
```diff
@@ -4,6 +4,8 @@
 from pathlib import Path
 import ast
 from collections import defaultdict
+import json
+import functools
```
Comment (BeadW) [inline dbt_invoke/properties.py:8]:
```suggestion
```

**Comment by BeadW** [comment]
@robastel nice suggestions. Let me know where you want to go from here. Perhaps we might want to whip up some more test examples if there are scenarios created here such as the one with the comments being left in a yaml file that you mentioned.

**Comment by robastel** [comment]
@BeadW If it works for you, I will merge this PR ([BeadW:main](https://github.com/BeadW/dbt-invoke) -> [Dashlane:combined-yml-migration](https://github.com/Dashlane/dbt-invoke/tree/combined-yml-migration)) and I can also of add a couple of tests (one for leftover comments and one to confirm that an empty file is successfully deleted) before I merge [Dashlane:combined-yml-migration](https://github.com/Dashlane/dbt-invoke/tree/combined-yml-migration) into [Dashlane:main](https://github.com/Dashlane/dbt-invoke/tree/main).

**Comment by BeadW** [comment]
Sounds great @robastel really appreciate you working with me on this one.

**Comment by robastel** [comment]
Same here!
